diffkemp/simpll/passes/FieldAccessFunctionGenerator.cpp

```diff
@@ -0,0 +1,191 @@
+//= FieldAccessFunctionGenerator.cpp - Move field access blocks to function ==//
+//
+//       SimpLL - Program simplifier for analysis of semantic difference      //
+//
+// This file is published under Apache 2.0 license. See LICENSE for details.
+// Author: Tomas Glozar, tglozar@gmail.com
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the definition of the FieldAccessFunctionGenerator pass.
+///
+//===----------------------------------------------------------------------===//
+
+#include "FieldAccessFunctionGenerator.h"
+
+#include "CalledFunctionsAnalysis.h"
+#include "Utils.h"
+#include <llvm/IR/Instructions.h>
+
+PreservedAnalyses FieldAccessFunctionGenerator::run(
+        Module &Mod,
+        AnalysisManager<Module, Function *> &mam,
+        Function *Main,
+        Module *ModOther) {
+    // This pass (similarly to RemoveUnusedReturnValuesPass, with which it
+    // shares the same template) is only performed on functions that are called
+    // somewhere in the compared code. Also although this is a pass on
+    // functions, it has to be implemented as a module pass, because it adds new
+    // functions to the module.
+    auto &CalledFuns = mam.getResult<CalledFunctionsAnalysis>(Mod, Main);
+
+    for (Function &Fun : Mod) {
+        if (!ModOther->getFunction(Fun.getName()))
+            continue;
+
+        if (CalledFuns.find(&Fun) == CalledFuns.end())
+            continue;
+
+        std::vector<Instruction *> InstStack;
+
+        // This stores the debug location for the currently processed stack.
+        // This enables the stopping of the building of the stack when an
+        // instruction with a different DILocation is reached.
+        DILocation *DIL;
+
+        // This is a FSM-like algorithm that gathers together groups of
+        // instructions beginning with a GEP and containing GEPs and casts.
+        // (They also have to have the same debug location, i.e. correspond to
+        // a single field access in the C source code).
+        // A vector/stack is generated from each of them and the processStack
+        // function is called to further process the instructions.
+        for (auto &BB : Fun) {
+            for (auto &Inst : BB) {
+                if (isa<GetElementPtrInst>(Inst) && InstStack.empty()) {
+                    // Possible start of field access block.
+                    // Check if GEP has constant indices (this pass could
+                    // technically support all GEPs, but this is not implemented
+                    // yet)
+                    if (dyn_cast<GetElementPtrInst>(&Inst)->
+                            hasAllConstantIndices()) {
+                        DIL = Inst.getDebugLoc().get();
```

Comment:
Since this is the only possible start of a field access block, you can clear `Stack` here.

Reply:
I don't think I can do that, because the algorithm uses the emptiness of the stack as an indication whether it is inside or outside a block.

Reply:
Right, then you can clear it after calling `processStack`. Or you can create a wrapper function that would call `processStack` and then it would clean the stack.

Reply:
Maybe, the best would be to name the wrapper function `processStack` and rename the current `processStack` to something like `createFunctionFromStack`.

Reply:
I think that creating a wrapper function just for that is not necessary, there are not that many places where the stack is processed, clearing it explicitly after each call should be fine.

Reply:
Ok, however, try to think if it is possible that we will do something more with the stack in future that could extend this wrapper function. If no, then stick with clearing the stack after each call. If it is possible that we do something more with the stack or we call `processStack` from other locations, then create a wrapper function.